### ISSUE TITLE
Self-training with weak labels

### ISSUE THREAD

**Issue by KrishnanJothi** [yueyu1030/COSINE#24]
Hi,

Thanks for publishing the paper and the code. 

I am using the code for weakly-supervised learning. I am using this code on a private dataset for token classification. 
 
I have these following questions.

1, I see that the mitr dataset is doing token classification. But in the file I am unable to find the [processor](https://github.com/yueyu1030/COSINE/blob/e3bb337a20cf830b9f0e07e3a06eeae48d8983cb/data_loader_new.py#L454) related to mitr in [data_loader_new.py](https://github.com/yueyu1030/COSINE/blob/e3bb337a20cf830b9f0e07e3a06eeae48d8983cb/data_loader_new.py)

2, In the paper it is mentioned that during the initialization with weak-labels, early stopping is adopted to avoid overfitting. But in the [trainer.py](https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/trainer.py), I am unable to find the part of the code which is responsible for early stopping.

3, I think in the [selftrain()](https://github.com/yueyu1030/COSINE/blob/e3bb337a20cf830b9f0e07e3a06eeae48d8983cb/trainer.py#L126), the [loss](https://github.com/yueyu1030/COSINE/blob/e3bb337a20cf830b9f0e07e3a06eeae48d8983cb/trainer.py#L199) is only the sum of prediction/classification loss and the contrastive loss. But in the paper it is mentioned that the confidence regularization loss is also a part of the loss function.

4, I think the [trainer.save_features()](        https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/main.py#L38) function is not defined in the [trainer.py](https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/trainer.py)

5, And also from the old issue, I too have the same questions from this [issue](https://github.com/yueyu1030/COSINE/issues/20#issuecomment-1003223064)

Kindly help me out or give some suggestions regarding my queries.

Thanks,
Krishnan

**Comment by yueyu1030** [comment]
Hi Krishnan,

Thanks for reaching out. Here is some of my explanations: 
1) This code repo does not contain the script for token classification. If you need this code, please let me know about your email address so I can send you a copy.
2) You can set the steps here https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/trainer.py#L243 for early stopping.
3) Contrastive loss is defined here https://github.com/yueyu1030/COSINE/blob/e3bb337a20cf830b9f0e07e3a06eeae48d8983cb/trainer.py#L78.
4) You can simply ignore this function right now, this function is not used in training.
5) Unfortunately I fail to understand what this issue means. Seems that this question is for semi-supervised learning, not weakly supervised learning.

Besides, a kind reminder is that our code has been integrated into the Wrench benchmark (https://github.com/JieyuZ2/wrench). The implementation can be found at https://github.com/JieyuZ2/wrench/blob/main/wrench/endmodel/cosine.py#L66. The codebase is cleaner and can be easily used. You may also consider using that codebase. Hope these explanations help.

Best,
Yue

**Comment by KrishnanJothi** [comment]
Hi Yue,

Thanks a lot for your reply. I will share you my email in Linkedin. 

I think my 3rd question was not clear. Can you please point out the line of code for "confidence regularization loss (R2 in the paper)"?

**Comment by yueyu1030** [comment]
Hi Krishnan, 

The reweighting term is in https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/trainer.py#L67; 
The confidence term is in https://github.com/yueyu1030/COSINE/blob/174d270a66adff7e8eebdd7bba2eb3c4884ce48a/trainer.py#L75.

Let me know if you have other questions.

Thanks,
Yue
